src/dune_rules/pkg_rules.ml

```diff
@@ -1757,9 +1757,27 @@ let build_rule context_name ~source_deps (pkg : Pkg.t) =
         in
         copy_action
         @ List.map pkg.info.extra_sources ~f:(fun (local, _) ->
+          (* If the package has extra sources, they will be
+             initially stored in the extra_sources directory for that
+             package. Prior to building, the contents of
+             extra_sources must be copied into the package's source
+             directory. *)
           let src = Paths.extra_source pkg.paths local in
           let dst = Path.Build.append_local pkg.write_paths.source_dir local in
-          Action.copy src dst |> Action.Full.make |> Action_builder.With_targets.return)
+          Action.progn
+            [ (* If the package has no source directory (some
+                 low-level packages are exclusively made up of extra
+                 sources), the source directory is first created. *)
+              Action.mkdir pkg.write_paths.source_dir
+            ; (* It's possible for some extra sources to already be at
+                 the destination. If these files are write-protected
+                 then the copy action will fail if we don't first remove
+                 them. *)
+              Action.remove_tree dst
```

Comment:
Why are removing existing sources though? Shouldn't we merge all the sources?

Reply:
that confused me as well, apparently this is to encode the semantics that extra-sources override the contents of the tarball.
I think a dedicated test for this would be useful.
In addition, maybe we can skip unpacking files present in `extra_sources` and skip this `remove_tree`, or something like that?

Reply:
If the source tarball has a file (or directory) and the extra-source has the same path I would expect that the extra-source would replace the original file as it is meant to patch the source tarball.

Reply:
Ok let's just make sure we have a test for this and we are matching opam's behavior.

Reply:
Tests are added in https://github.com/ocaml/dune/pull/10790. I updated that PR with a test for the scenario where an extra-source file replaces a file in the source dir and did a manual experiment to confirm that this is also how opam behaves.